fdbserver/include/fdbserver/GrvTransactionRateInfo.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * GrvTransactionRateInfo.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "fdbrpc/Smoother.h"
+
+// Used by GRV Proxy to enforce rate limits received from the Ratekeeper.
+//
+// Between delays, the GrvTransactionRateInfo executes an "epoch" starting
+// with a call to the startEpoch method. Within this epoch, transactions are
+// released while canStart returns true. At the end of the epoch, the
+// endEpoch method is called, and the budget is updated to add or
+// remove  capacity.
+//
+// Meanwhile, the desired rate is updated through the setRate method.
+//
+// Smoothers are used to avoid turbulent throttling behaviour.
+class GrvTransactionRateInfo {
+	double rate{ 0.0 };
+	double limit{ 0.0 };
+	double budget{ 0.0 };
+	bool disabled{ true };
+	Smoother smoothRate;
+	Smoother smoothReleased;
+
+public:
+	explicit GrvTransactionRateInfo(double rate = 0.0);
+
+	// Determines the number of transactions that this proxy is allowed to release
+	// in this epoch.
+	void startEpoch();
```

Comment:
I think referring to these as epochs should be ok, but we may want to carefully define what an epoch is. The `GrvTransactionRateInfo` is conceptually using a sliding window and trying to keep the number of transaction started in that window consistent with the limit, and this is of course distinct from the epoch. I could see that potentially causing confusion, and some explanation may help with that.

Reply:
I renamed "epoch" to "releaseWindow" to avoid confusion.